packages/@lwc/template-compiler/src/parser/expression-complex/html.ts

```diff
@@ -0,0 +1,177 @@
+const WHITESPACE = /(\s*)/;
+
+function getWhitespaceLen(str: string): number {
+    return WHITESPACE.exec(str)![0].length;
+}
+
+/**
+ * This class extends `parse5`'s internal tokenizer.
+ *
+ * Its behavior diverges from that specified in the WHATWG HTML spec
+ * in two places:
+ *   - 13.2.5.38 - unquoted attribute values
+ *   - 13.2.5.1 - the "data" state, which corresponds to parsing outside of tags
+ *
+ * Specifically, this tokenizer defers to Acorn's JavaScript parser when
+ * encountering a `{` character for an attribute value or within a text
+ * node. Acorn parses the expression, and the tokenizer continues its work
+ * following the closing `}`.
+ */
+class TemplateHtmlTokenizer extends Tokenizer {
+    parser: TemplateHtmlParser;
+    // We track which attribute values are in-progess so that we can defer
+    // to the default tokenizer's behavior after the first character of
+    // an unquoted attr value has been checked for an opening curly brace.
+    checkedAttrs = new WeakSet<any>();
+
+    constructor(parser: TemplateHtmlParser) {
+        super();
+        this.parser = parser;
+    }
+
+    parseTemplateExpression() {
+        const expressionStart: number = this.preprocessor.pos;
+        const html = this.preprocessor.html as string;
+
+        const leadingWhitespaceLen = getWhitespaceLen(html.slice(expressionStart + 1));
+        const javascriptExprStart = expressionStart + leadingWhitespaceLen + 1;
+
+        // Start parsing immediately after the opening curly brace.
+        const estreeNode = parseExpressionAt(html, javascriptExprStart, {
+            ecmaVersion: 2022,
+            locations: true,
+            ranges: true,
+        });
+
+        const javascriptExprLen = estreeNode.end - estreeNode.start;
+        const javascriptExprEnd = javascriptExprStart + javascriptExprLen;
+        const trailingWhitespaceLen = getWhitespaceLen(html.slice(javascriptExprEnd));
+        const idxOfClosingBracket = javascriptExprEnd + trailingWhitespaceLen;
+        // Capture text content between the outer curly braces, inclusive.
+        const expressionTextNodeValue = html.slice(expressionStart, idxOfClosingBracket + 1);
+
+        invariant(
+            html.codePointAt(idxOfClosingBracket) === CLOSING_CURLY_BRACKET,
+            ParserDiagnostics.TEMPLATE_EXPRESSION_PARSING_ERROR,
+            ['expression must end with curly brace.']
+        );
+
+        // The lexer's cursor (pos) will be advanced by one after we conclude
+        // parsing of the template expression, so `pos` is left pointing to the
+        // closing `}` character.
+        this.preprocessor.pos = idxOfClosingBracket;
+
+        // Parsed expressions that are cached here will be later retrieved when the
+        // LWC template AST is being constructed.
+        this.parser.preparsedJsExpressions.set(expressionStart, estreeNode);
+
+        return expressionTextNodeValue;
+    }
+
+    ATTRIBUTE_VALUE_UNQUOTED_STATE(cp: number) {
```

Comment:
Nit: I think `codePoint` is clearer than `cp`.

Reply:
I reused the variable that `parse5` uses internally, but I agree that readability suffers. I'll make the change.